Adding gh release download for .zip and .tar.gz

lpessoa opened the pull request:
Fixes #2167

Adding ```gh release download``` support for fetching the default .zip or .tar.gz release files.

The user will be able to require the files by simply adding the ```--archive``` command line parameter.
```bash
gh release download v1.3.0 --archive=tgz
```

####
- [X] Implement basic functionality
- [x] Add stricter argument validations
- [x] Create tests

<!--
  Thank you for contributing to GitHub CLI!
  To reference an open issue, please write this in your description: `Fixes #NUMBER`
-->

lpessoa committed:
Adding gh release download for .zip and .tar.gz

Co-authored-by: Mislav Marohnić <mislav@github.com>

mislav commented on pkg/cmd/release/download/download.go:
```diff
@@ -47,12 +51,15 @@ func NewCmdDownload(f *cmdutil.Factory, runF func(*DownloadOptions) error) *cobr
 			# specify multiple file patterns
 			$ gh release download -p '*.deb' -p '*.rpm'
+
+			# download default zip package from a specific release
```
Instead of saying `default * package` in the command docs, let's be consistent with the web UI where these archives are called "source code". How would this sound:
```suggestion
			# download the archive of the source code for a release
```

mislav commented on pkg/cmd/release/download/download.go:
```diff
@@ -78,10 +90,29 @@ func NewCmdDownload(f *cmdutil.Factory, runF func(*DownloadOptions) error) *cobr
 
 	cmd.Flags().StringVarP(&opts.Destination, "dir", "D", ".", "The directory to download files into")
 	cmd.Flags().StringArrayVarP(&opts.FilePatterns, "pattern", "p", nil, "Download only assets that match a glob pattern")
+	cmd.Flags().StringVarP(&opts.ArchiveType, "archive", "A", "", "Download default assset with specified type (zip or tgz)")
```
1. For consistency with the web interface, let's use `tar.gz` instead of `tgz` since the former is the extension you would end up with if you've chosen to download a "tar.gz" from a release.
2. Any reason you preferred `-A` instead of `-a` as shorthand? I think either would be fine, but I'm interested in your thoughts.
3. Lets use `` `format` `` enclosed in backticks in the flag description because that renders nicely in the flag help docs.
```suggestion
	cmd.Flags().StringVarP(&opts.ArchiveType, "archive", "A", "", "Download the source code archive in the specified `format` (zip or tar.gz)")
```

lpessoa replied:
Used the ```-A``` just because some tools i've used before tend to use ```-a``` for something related to _all_ 😄 .
But we can go with the lowercase option if it is more consistent with the CLI pattern.

mislav commented on pkg/cmd/release/download/download.go:
```diff
@@ -78,10 +90,29 @@ func NewCmdDownload(f *cmdutil.Factory, runF func(*DownloadOptions) error) *cobr
+		"zip": {},
+		"tgz": {},
+	}
+	if _, ok := validArchiveTypes[opts.ArchiveType]; !ok {
```
This is neat and pretty idiomatic Go, but we're unlikely to ever add more formats to this list, so let's keep it a simple conditional to avoid allocating a map:
```suggestion
	if opts.ArchiveType != "zip" && opts.ArchiveType != "tar.gz" {
```

mislav commented on pkg/cmd/release/download/download.go:
```diff
@@ -78,10 +90,29 @@ func NewCmdDownload(f *cmdutil.Factory, runF func(*DownloadOptions) error) *cobr
+	}
+
+	if len(opts.FilePatterns) > 0 {
+		return &cmdutil.FlagError{Err: errors.New("the '--pattern' and '--archive' flags cannot be set simultaneously")}
```
We have an idiomatic helper for this: cmdutil.MutuallyExclusive

mislav commented on pkg/cmd/release/download/download.go:
```diff
@@ -179,13 +224,16 @@ func downloadAssets(httpClient *http.Client, toDownload []shared.ReleaseAsset, d
 
 	req.Header.Set("Accept", "application/octet-stream")
+	if len(archiveType) > 0 {
+		req.Header.Set("Accept", "application/octet-stream, application/json")
```
Why include `application/json` here? Does it change anything? If we didn't need it then the `downloadAsset` function doesn't need to accept an extra argument

lpessoa replied:
During my tests with the actual API endpoint i would get an error from the server regarding the client not accepting the application/json in the ```Accept``` request header. No idea why actually.. have you seen it before @mislav ?

mislav replied:
Gah, I can see you are right. Definitely a bug with the API endpoint. Your workaround seems fine, but please:
1. Add a code comment above the workaround explaining why this is needed;
2. Change the `archiveType` function argument to `isArchive bool`.

mislav commented on pkg/cmd/release/download/download.go:
```diff
@@ -108,11 +139,25 @@ func downloadRun(opts *DownloadOptions) error {
+		}
+
+		asset := &shared.ReleaseAsset{
+			Name:   fmt.Sprintf("%s.%s", opts.TagName, opts.ArchiveType),
```
The `Name` field is only used to name the local file that the asset will be downloaded to, and it would be great if this was consistent with how the tarball/zipball endpoint would normally name the file. Typically, the name would be something like `{reponame}-{tagname}.{ext}`, but we don't have to guess or generate this name ourselves. Let's just let the server decide on the local name for this asset! I propose to leave the `Name` field blank and instead populate the name of the local file based on the `content-disposition: attachment; filename=MYFILE.EXT` response header from the server.

lpessoa replied:
Makes sense on using the reply header definition 👍🏼 
I just have to change the existing ```downloadAsset``` method to use the reply header value if `Name` is not set in the `ReleaseAsset` struct

mislav commented on pkg/cmd/release/download/download.go:
```diff
@@ -108,11 +139,25 @@ func downloadRun(opts *DownloadOptions) error {
-			continue
+	// create Asset with ZipBallURL or TarBallURL if using archive option
+	if len(opts.ArchiveType) > 0 {
+		var defaultUrl = release.ZipballURL
```
Let's call this archiveURL:
```suggestion
		archiveURL := release.ZipballURL
```

mislav reviewed (changes requested):
Thank you for working on this; this is shaping up great!

A general note about ZipballURL/TarballURL API endpoints: this might not be immediately apparent, but these APIs result in a "legacy" format of the zip/tgz archive when compared to the archives downloaded by using the "Source code" download links from the web interface. Ideally this would be consistent with the web, as I personally don't like the legacy archive format and find the non-legacy format way more useful. However, you don't need to address this in your PR; I'm asking other people at GitHub more knowledgeable about Releases as to what we could potentially do here.

lpessoa commented on pkg/cmd/release/download/download.go:
```diff
@@ -241,6 +246,23 @@ func downloadAsset(httpClient *http.Client, assetURL, destinationPath string, ar
+
+		_, params, err := mime.ParseMediaType(contentDisposition)
+		if err != nil {
+			return err
```
@mislav Just returning the error here. WDYT of using the tag name in case the content-disposition header is missing?

mislav replied:
No, I think it's fine to error out. Maybe wrap the error like:
```suggestion
			return fmt.Errorf("unable to parse file name of archive: %w", err)
```

lpessoa reviewed (commented):
Code review changes implemented. Added a little helper to httpmock to allow setting a Header value for the Responder's http.response (can be easilly converted to accept multiple headers).
Just a small question regarding the scenario where the content-disposition header is missing from the server (should not happen 😄 )

mislav commented on pkg/cmd/release/download/download.go:
```diff
@@ -141,14 +143,13 @@ func downloadRun(opts *DownloadOptions) error {
-		if strings.ToLower(opts.ArchiveType) == "tgz" {
-			defaultUrl = release.TarballURL
+		var archiveURL = release.ZipballURL
+		if strings.ToLower(opts.ArchiveType) == "tar.gz" {
```
Currently it's not possible to pass in `TAR.GZ` (uppercase) due to flag validation, so therefore I don't think it's necessary to `strings.ToLower()` here.

mislav commented on pkg/cmd/release/download/download.go:
```diff
@@ -245,6 +246,23 @@ func downloadAsset(httpClient *http.Client, assetURL, destinationPath string, ar
+		}
+	}
+
+	fmt.Println(destinationDir)
```
This should go away

mislav commented on pkg/cmd/release/download/download.go:
```diff
@@ -245,6 +246,23 @@ func downloadAsset(httpClient *http.Client, assetURL, destinationPath string, ar
+		if serverFileName, ok := params["filename"]; ok {
+			destinationPath = filepath.Join(destinationDir, serverFileName)
+		} else {
+			return fmt.Errorf("unable to fetch file name from response's Content-Disposition header")
```
Nitpick: this error message leaks too much about the implementation (the user doesn't care about a "Content-Disposition" header). Maybe something like:
```suggestion
			return errors.New("unable to determine file name of archive")
```

mislav reviewed (changes requested):
Looks great! Only minor bits remaining

mislav reviewed (approved):
Looks great!

I will hold off merging until we internally resolve the question of how do we access non-legacy zipball/tarball archives (see my previous review) using the public API.

shiouen commented:
awesome @lpessoa!

@mislav any update on that internal question?

mislav commented:
@shiouen No movement from the platform side. It looks like we might be stuck with the currently available API that redirects to legacy tarball/zipball format. However, we can take control over the HTTP redirect and remove the `legacy.` bit from the codeload URL manually in order to access non-legacy formats. It's a hack, but I think it's a worthwhile one. I'll try to add it to the PR!

mislav committed:
Ensure that we never download "legacy" archive formats

The `zipball_url` and `tarball_url` fields on the Release API payload
always link to "legacy" archives that are of a different format than
those one would get if they downloaded an archive from a Release using
the web interface. The GitHub API does not seem to publish links to
non-legacy archives.

This adds a redirect hack to turn "legacy" Codeload URLs into non-legacy
ones with the goal of ensuring the consistency of user experience when
downloading archives.

mislav committed:
Allow downloading the archive of the latest release
